mempool/src/pool.rs

```diff
@@ -790,6 +801,25 @@ where
             .collect()
     }
 
+    fn collect_txs(
+        &self,
+        mut tx_accumulator: Box<dyn TransactionAccumulator>,
+    ) -> Vec<SignedTransaction> {
+        if self.store.is_empty() {
+            return Vec::new();
+        }
+        let mut tx_iter = self.store.txs_by_descendant_score.values().flatten();
+        while !tx_accumulator.done() {
+            if let Some(next_tx_id) = tx_iter.next() {
+                let next_tx = self.store.txs_by_id.get(next_tx_id).expect("tx to exist");
+                tx_accumulator.add_tx(next_tx.tx().clone());
+            } else {
```

Comment:
Shouldn't we pop transactions from the mempool when they're collected?

Reply:
I'm not sure. Bitcoin Core drops transactions from the mempool when a new tip is connected, not when collecting txs for a block.
I can imagine the following scenario: we collect transactions for a block, try to mine it, and before we succeed another, competing block arrives over the network. In this case if we have dropped the txs from the mempool during our attempted block production, those txs will be lost forever.